File: genium/src/main/java/com/here/genium/generator/java/JavaModelBuilder.java
/*
 * Copyright (c) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.ivi.api.generator.java;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Iterables;
import com.here.ivi.api.common.CollectionsHelper;
import com.here.ivi.api.common.FrancaTypeHelper;
import com.here.ivi.api.generator.common.modelbuilder.AbstractModelBuilder;
import com.here.ivi.api.generator.common.modelbuilder.ModelBuilderContextStack;
import com.here.ivi.api.model.franca.CommentHelper;
import com.here.ivi.api.model.franca.FrancaDeploymentModel;
import com.here.ivi.api.model.java.*;
import com.here.ivi.api.model.java.JavaMethod.MethodQualifier;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.franca.core.franca.*;

public class JavaModelBuilder extends AbstractModelBuilder<JavaElement> {

  /**
   * Used to store a unique JavaExceptionClass element of every method that throws. They will be
   * later generated in a dedicated Java file from JavaGenerator
   */
  public final Map<String, JavaExceptionClass> exceptionClasses = new HashMap<>();

  private final FrancaDeploymentModel deploymentModel;
  private final JavaPackage rootPackage;
  private final JavaTypeMapper typeMapper;
  private final JavaType nativeBase;

  @VisibleForTesting
  JavaModelBuilder(
      final ModelBuilderContextStack<JavaElement> contextStack,
      final FrancaDeploymentModel deploymentModel,
      final JavaPackage rootPackage,
      final JavaTypeMapper typeMapper) {
    super(contextStack);
    this.deploymentModel = deploymentModel;
    this.rootPackage = rootPackage;
    this.typeMapper = typeMapper;
    this.nativeBase = typeMapper.getNativeBase();
  }

  public JavaModelBuilder(
      final FrancaDeploymentModel deploymentModel,
      final JavaPackage rootPackage,
      final JavaTypeMapper typeMapper) {
    this(new ModelBuilderContextStack<>(), deploymentModel, rootPackage, typeMapper);
  }

  @Override
  public void finishBuilding(FInterface francaInterface) {

    if (deploymentModel.isInterface(francaInterface)) {
      finishBuildingFrancaInterface(francaInterface);
    } else {
      finishBuildingFrancaClass(francaInterface);
    }

    closeContext();
  }

  @Override
  public void finishBuilding(FTypeCollection francaTypeCollection) {

    CollectionsHelper.getStreamOfType(getCurrentContext().previousResults, JavaClass.class)
        .forEach(this::storeResult);
    CollectionsHelper.getStreamOfType(getCurrentContext().previousResults, JavaEnum.class)
        .forEach(this::storeResult);
    closeContext();
  }

  @Override
  public void finishBuilding(FMethod francaMethod) {

    boolean needSelectorSuffix =
        francaMethod.getSelector() != null && FrancaTypeHelper.hasArrayParameters(francaMethod);
    String selector = needSelectorSuffix ? francaMethod.getSelector() : "";
    String javaMethodName = JavaNameRules.getMethodName(francaMethod.getName(), selector);

    // Map return type
    JavaParameter outputParameter =
        CollectionsHelper.getStreamOfType(getCurrentContext().previousResults, JavaParameter.class)
            .filter(parameter -> parameter.isOutput)
            .findFirst()
            .orElse(null);

    JavaType returnType;
    String returnComment;
    if (outputParameter == null) { // Void return type
      returnType = JavaPrimitiveType.VOID;
      returnComment = null;
    } else {
      returnType = outputParameter.type;
      returnComment = outputParameter.comment;
    }

    JavaCustomType javaExceptionTypeRef = null;
    JavaEnumType exceptionEnumTypeRef = getPreviousResult(JavaEnumType.class);
    if (exceptionEnumTypeRef != null) {

      String exceptionName =
          JavaNameRules.getExceptionName(Iterables.getLast(exceptionEnumTypeRef.classNames));
      JavaPackage exceptionPackage = new JavaPackage(exceptionEnumTypeRef.packageNames);
      JavaExceptionClass javaExceptionClass =
          new JavaExceptionClass(exceptionName, exceptionEnumTypeRef, exceptionPackage);
      String mapKey = exceptionPackage.packageNames + "." + exceptionName;
      exceptionClasses.put(mapKey, javaExceptionClass);

      javaExceptionTypeRef = new JavaCustomType(exceptionName, exceptionPackage);
    }

    JavaMethod javaMethod =
        JavaMethod.builder(javaMethodName)
            .returnType(returnType)
            .returnComment(returnComment)
            .exception(javaExceptionTypeRef)
            .build();
    javaMethod.comment = CommentHelper.getDescription(francaMethod);

    if (deploymentModel.isStatic(francaMethod)) {
      javaMethod.qualifiers.add(MethodQualifier.STATIC);
    }
    javaMethod.visibility = getVisibility(francaMethod);

    List<JavaParameter> inputParameters =
        CollectionsHelper.getStreamOfType(getCurrentContext().previousResults, JavaParameter.class)
            .filter(parameter -> !parameter.isOutput)
            .collect(Collectors.toList());
    javaMethod.parameters.addAll(inputParameters);

    storeResult(javaMethod);
    closeContext();
  }

  @Override
  public void finishBuildingInputArgument(FArgument francaArgument) {

    JavaType javaArgumentType = getPreviousResult(JavaType.class);

    JavaParameter javaParameter =
        new JavaParameter(
            javaArgumentType, JavaNameRules.getArgumentName(francaArgument.getName()));
    javaParameter.comment = CommentHelper.getDescription(francaArgument);

    storeResult(javaParameter);
    closeContext();
  }

  @Override
  public void finishBuildingOutputArgument(FArgument francaArgument) {

    JavaType javaArgumentType =
        CollectionsHelper.getFirstOfType(getCurrentContext().previousResults, JavaType.class);

    JavaParameter javaParameter =
        new JavaParameter(
            javaArgumentType, JavaNameRules.getArgumentName(francaArgument.getName()), true);
    javaParameter.comment = CommentHelper.getDescription(francaArgument);

    storeResult(javaParameter);
    closeContext();
  }

  @Override
  public void finishBuilding(FConstantDef francaConstant) {

    JavaType javaType =
        CollectionsHelper.getFirstOfType(getCurrentContext().previousResults, JavaType.class);
    JavaValue value = JavaValueMapper.map(javaType, francaConstant.getRhs());
    JavaConstant javaConstant =
        new JavaConstant(javaType, JavaNameRules.getConstantName(francaConstant.getName()), value);

    javaConstant.comment = CommentHelper.getDescription(francaConstant);

    storeResult(javaConstant);
    closeContext();
  }

  @Override
  public void finishBuilding(FField francaField) {

    String fieldName = JavaNameRules.getFieldName(francaField.getName());
    JavaType javaType = getPreviousResult(JavaType.class);
    String defaultValue = deploymentModel.getDefaultValue(francaField);
    JavaValue initialValue =
        defaultValue != null
            ? JavaValueMapper.mapDefaultValue(javaType, defaultValue)
            : JavaValueMapper.mapDefaultValue(javaType);
    boolean isNonNull = deploymentModel.isNotNull(francaField);

    JavaField javaField =
        JavaField.builder(fieldName, javaType).initial(initialValue).isNonNull(isNonNull).build();
    javaField.visibility = getVisibility(francaField);
    javaField.comment = CommentHelper.getDescription(francaField);

    if (isNonNull) {
      JavaType notNullAnnotation = typeMapper.getNotNullAnnotation();
      if (notNullAnnotation != null) {
        javaField.annotations.add(notNullAnnotation);
      }
    }

    storeResult(javaField);
    closeContext();
  }

  @Override
  public void finishBuilding(FStructType francaStructType) {

    JavaType serializationBase = typeMapper.getSerializationBase();
    boolean isSerializable =
        serializationBase != null && deploymentModel.isSerializable(francaStructType);

    // Type definition
    JavaClass javaClass =
        JavaClass.builder(JavaNameRules.getClassName(francaStructType.getName()))
            .extendedClass(getPreviousResult(JavaCustomType.class))
            .isParcelable(isSerializable)
            .build();
    javaClass.visibility = getVisibility(francaStructType);
    javaClass.javaPackage = rootPackage;

    javaClass.comment = CommentHelper.getDescription(francaStructType);
    javaClass.fields.addAll(getPreviousResults(JavaField.class));

    JavaClass parentClass = getPreviousResult(JavaClass.class);
    if (parentClass != null) {
      javaClass.parentFields.addAll(parentClass.parentFields);
      javaClass.parentFields.addAll(parentClass.fields);
    }

    if (isSerializable) {
      javaClass.parentInterfaces.add(serializationBase);
    }

    storeResult(javaClass);

    // Type reference
    storeResult(typeMapper.mapCustomType(francaStructType));

    closeContext();
  }

  @Override
  public void finishBuilding(FTypeRef francaTypeRef) {
    storeResult(typeMapper.map(francaTypeRef));
    closeContext();
  }

  @Override
  public void finishBuilding(FEnumerationType francaEnumType) {

    // Type definition
    JavaEnum javaEnum = new JavaEnum(JavaNameRules.getClassName(francaEnumType.getName()));
    javaEnum.visibility = getVisibility(francaEnumType);
    javaEnum.javaPackage = rootPackage;

    javaEnum.comment = CommentHelper.getDescription(francaEnumType);
    javaEnum.items.addAll(getPreviousResults(JavaEnumItem.class));
    JavaValueMapper.completePartialEnumeratorValues(javaEnum.items);
    storeResult(javaEnum);

    // Type reference
    storeResult(typeMapper.mapCustomType(francaEnumType));

    closeContext();
  }

  @Override
  public void finishBuilding(FExpression francaExpression) {
    storeResult(JavaValueMapper.map(francaExpression));
    closeContext();
  }

  @Override
  public void finishBuilding(FEnumerator francaEnumerator) {

    String enumItemName = JavaNameRules.getConstantName(francaEnumerator.getName());
    JavaValue javaValue = getPreviousResult(JavaValue.class);
    JavaEnumItem javaEnumItem = new JavaEnumItem(enumItemName, javaValue);

    javaEnumItem.comment = CommentHelper.getDescription(francaEnumerator);

    storeResult(javaEnumItem);
    closeContext();
  }

  @Override
  public void finishBuilding(FArrayType francaArrayType) {

    storeResult(typeMapper.mapArray(francaArrayType));
    closeContext();
  }

  @Override
  public void finishBuilding(FMapType francaMapType) {

    storeResult(typeMapper.mapMap(francaMapType));
    closeContext();
  }

  @Override
  public void finishBuilding(FAttribute francaAttribute) {

    JavaType javaType = getPreviousResult(JavaType.class);
    String comment = CommentHelper.getDescription(francaAttribute);

    JavaVisibility visibility = getVisibility(francaAttribute);

    String getterName = JavaNameRules.getGetterName(francaAttribute.getName(), javaType);

    JavaMethod getterMethod = JavaMethod.builder(getterName).returnType(javaType).build();
    getterMethod.visibility = visibility;
    getterMethod.comment = comment;

    storeResult(getterMethod);

    if (!francaAttribute.isReadonly()) {
      String setterName = JavaNameRules.getSetterName(francaAttribute.getName());

      JavaMethod setterMethod = JavaMethod.builder(setterName).build();
      setterMethod.visibility = visibility;
      setterMethod.comment = comment;

      setterMethod.parameters.add(new JavaParameter(javaType, "value"));
      storeResult(setterMethod);
    }

    closeContext();
  }

  private JavaClass createJavaClass(
      final FInterface francaInterface,
      final List<JavaMethod> methods,
      final JavaType extendedClass) {

    JavaClass javaClass =
        JavaClass.builder(JavaNameRules.getClassName(francaInterface.getName()))
            .extendedClass(extendedClass)
            .isImplClass(true)
            .needsDisposer(nativeBase.equals(extendedClass))
            .build();
    javaClass.visibility = getVisibility(francaInterface);
    javaClass.javaPackage = rootPackage;

    javaClass.comment = CommentHelper.getDescription(francaInterface);
    javaClass.fields.addAll(getPreviousResults(JavaField.class));

    javaClass.constants.addAll(getPreviousResults(JavaConstant.class));
    javaClass.methods.addAll(methods);
    javaClass.methods.forEach(method -> method.qualifiers.add(MethodQualifier.NATIVE));
    javaClass.enums.addAll(getPreviousResults(JavaEnum.class));

    addInnerClasses(javaClass);

    return javaClass;
  }

  private JavaInterface createJavaInterface(final FInterface francaInterface) {

    JavaInterface javaInterface =
        new JavaInterface(JavaNameRules.getClassName(francaInterface.getName()));
    javaInterface.visibility = getVisibility(francaInterface);
    javaInterface.javaPackage = rootPackage;

    javaInterface.comment = CommentHelper.getDescription(francaInterface);
    javaInterface.constants.addAll(getPreviousResults(JavaConstant.class));
    javaInterface.enums.addAll(getPreviousResults(JavaEnum.class));

    List<JavaMethod> interfaceMethods =
        getPreviousResults(JavaMethod.class)
            .stream()
            .map(javaMethod -> javaMethod.toBuilder().visibility(JavaVisibility.PACKAGE).build())
            .collect(Collectors.toList());
    javaInterface.methods.addAll(interfaceMethods);

    addInnerClasses(javaInterface);

    return javaInterface;
  }

  private JavaClass createJavaImplementationClass(
      final FInterface francaInterface,
      final JavaInterface javaInterface,
      final JavaType extendedClass) {

    JavaClass javaClass =
        JavaClass.builder(JavaNameRules.getImplementationClassName(francaInterface.getName()))
            .extendedClass(extendedClass)
            .isImplClass(true)
            .needsDisposer(nativeBase.equals(extendedClass))
            .build();
    javaClass.visibility = JavaVisibility.PACKAGE;
    javaClass.javaPackage = rootPackage;
    javaClass.fields.addAll(getPreviousResults(JavaField.class));

    javaClass.methods.addAll(getPreviousResults(JavaMethod.class));
    javaClass.methods.forEach(method -> method.qualifiers.add(MethodQualifier.NATIVE));

    JavaCustomType interfaceTypeReference =
        new JavaCustomType(javaInterface.name, javaInterface.javaPackage);
    javaClass.parentInterfaces.add(interfaceTypeReference);

    return javaClass;
  }

  private void addInnerClasses(final JavaTopLevelElement javaTopLevelElement) {
    List<JavaClass> innerClasses =
        getPreviousResults(JavaClass.class)
            .stream()
            .filter(javaClass -> !javaClass.isImplClass)
            .collect(Collectors.toList());
    innerClasses.forEach(innerClass -> innerClass.qualifiers.add(JavaClass.Qualifier.STATIC));
    javaTopLevelElement.innerClasses.addAll(innerClasses);
  }

  private void finishBuildingFrancaClass(final FInterface francaInterface) {

    JavaType extendedClass = nativeBase;

    FInterface parentInterface = francaInterface.getBase();
    if (parentInterface != null) {
      if (deploymentModel.isInterface(parentInterface)) {
        String parentImplementationClassName =
            JavaNameRules.getImplementationClassName(parentInterface.getName());
        extendedClass = typeMapper.mapCustomType(parentInterface, parentImplementationClassName);
      } else {
        extendedClass = typeMapper.mapCustomType(parentInterface);
      }
    }

    List<JavaMethod> methods = getPreviousResults(JavaMethod.class);
    JavaClass javaClass = createJavaClass(francaInterface, methods, extendedClass);

    storeResult(javaClass);
  }

  private void finishBuildingFrancaInterface(final FInterface francaInterface) {

    JavaInterface javaInterface = createJavaInterface(francaInterface);

    JavaType extendedClass = nativeBase;
    FInterface parentInterface = francaInterface.getBase();
    if (parentInterface != null) {
      javaInterface.parentInterfaces.add(typeMapper.mapCustomType(parentInterface));
      String parentImplementationClassName =
          JavaNameRules.getImplementationClassName(parentInterface.getName());
      extendedClass = typeMapper.mapCustomType(parentInterface, parentImplementationClassName);
    }

    JavaClass javaImplementationClass =
        createJavaImplementationClass(francaInterface, javaInterface, extendedClass);

    storeResult(javaInterface);
    storeResult(javaImplementationClass);
  }

  private JavaVisibility getVisibility(final FModelElement francaModelElement) {
    return deploymentModel.isInternal(francaModelElement)
        ? JavaVisibility.PACKAGE
        : JavaVisibility.PUBLIC;
  }
}

File: genium/src/main/java/com/here/genium/generator/java/JavaModelBuilder.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.java

import com.here.genium.common.ModelBuilderContextStack
import com.here.genium.generator.common.modelbuilder.AbstractLimeBasedModelBuilder
import com.here.genium.model.common.Comments
import com.here.genium.model.java.JavaClass
import com.here.genium.model.java.JavaConstant
import com.here.genium.model.java.JavaCustomType
import com.here.genium.model.java.JavaElement
import com.here.genium.model.java.JavaEnum
import com.here.genium.model.java.JavaEnumItem
import com.here.genium.model.java.JavaEnumType
import com.here.genium.model.java.JavaExceptionClass
import com.here.genium.model.java.JavaField
import com.here.genium.model.java.JavaInterface
import com.here.genium.model.java.JavaMethod
import com.here.genium.model.java.JavaMethod.MethodQualifier
import com.here.genium.model.java.JavaPackage
import com.here.genium.model.java.JavaParameter
import com.here.genium.model.java.JavaPrimitiveType
import com.here.genium.model.java.JavaTopLevelElement
import com.here.genium.model.java.JavaType
import com.here.genium.model.java.JavaValue
import com.here.genium.model.java.JavaVisibility
import com.here.genium.model.lime.LimeAttributeType
import com.here.genium.model.lime.LimeAttributeType.DEPRECATED
import com.here.genium.model.lime.LimeAttributeType.JAVA
import com.here.genium.model.lime.LimeAttributeValueType.BUILDER
import com.here.genium.model.lime.LimeAttributeValueType.MESSAGE
import com.here.genium.model.lime.LimeClass
import com.here.genium.model.lime.LimeConstant
import com.here.genium.model.lime.LimeContainerWithInheritance
import com.here.genium.model.lime.LimeEnumeration
import com.here.genium.model.lime.LimeEnumerator
import com.here.genium.model.lime.LimeException
import com.here.genium.model.lime.LimeField
import com.here.genium.model.lime.LimeFunction
import com.here.genium.model.lime.LimeInterface
import com.here.genium.model.lime.LimeNamedElement
import com.here.genium.model.lime.LimeParameter
import com.here.genium.model.lime.LimeProperty
import com.here.genium.model.lime.LimeStruct
import com.here.genium.model.lime.LimeTypeRef
import com.here.genium.model.lime.LimeTypesCollection
import com.here.genium.model.lime.LimeValue
import com.here.genium.model.lime.LimeVisibility
import java.util.EnumSet

class JavaModelBuilder(
    contextStack: ModelBuilderContextStack<JavaElement> = ModelBuilderContextStack(),
    private val rootPackage: JavaPackage,
    private val typeMapper: JavaTypeMapper,
    private val valueMapper: JavaValueMapper,
    private val nameRules: JavaNameRules
) : AbstractLimeBasedModelBuilder<JavaElement>(contextStack) {
    private val nativeBase: JavaType = typeMapper.nativeBase

    override fun finishBuilding(limeContainer: LimeContainerWithInheritance) {
        when (limeContainer) {
            is LimeInterface -> finishBuildingInterface(limeContainer)
            is LimeClass -> finishBuildingClass(limeContainer)
        }

        closeContext()
    }

    override fun finishBuilding(limeTypes: LimeTypesCollection) {
        getPreviousResults(JavaTopLevelElement::class.java).forEach { this.storeResult(it) }

        val constants = getPreviousResults(JavaConstant::class.java)
        if (constants.isNotEmpty()) {
            val javaClass = JavaClass(nameRules.getName(limeTypes))
            javaClass.visibility = getVisibility(limeTypes)
            javaClass.qualifiers.add(JavaTopLevelElement.Qualifier.FINAL)
            javaClass.javaPackage = rootPackage
            javaClass.comment = createComments(limeTypes)
            addDeprecatedAnnotationIfNeeded(javaClass)
            javaClass.constants.addAll(constants)

            storeNamedResult(limeTypes, javaClass)
        }

        closeContext()
    }

    override fun finishBuilding(limeMethod: LimeFunction) {
        val returnType = when {
            limeMethod.isConstructor -> {
                val parentType = typeMapper.mapParentType(limeMethod) as JavaCustomType
                if (parentType.isInterface) JavaPrimitiveType.LONG else parentType
            }
            else -> {
                val cppType = typeMapper.mapType(limeMethod.returnType.typeRef)
                typeMapper.applyNullability(cppType, limeMethod.returnType.typeRef.isNullable)
            }
        }

        val javaExceptionType = limeMethod.thrownType?.let { typeMapper.mapExceptionType(it) }
        val qualifiers = when {
            limeMethod.isStatic -> setOf(MethodQualifier.STATIC)
            else -> emptySet()
        }
        val javaMethod = JavaMethod(
            name = nameRules.getName(limeMethod),
            comment = createComments(limeMethod),
            visibility = getVisibility(limeMethod),
            returnType = returnType,
            returnComment = limeMethod.returnType.comment.getFor(PLATFORM_TAG),
            exception = javaExceptionType,
            throwsComment = limeMethod.thrownType?.comment?.getFor(PLATFORM_TAG),
            parameters = getPreviousResults(JavaParameter::class.java),
            isConstructor = limeMethod.isConstructor,
            qualifiers = qualifiers
        )
        addDeprecatedAnnotationIfNeeded(javaMethod)

        storeNamedResult(limeMethod, javaMethod)
        closeContext()
    }

    override fun finishBuilding(limeParameter: LimeParameter) {
        val javaType = typeMapper.applyNullability(
            getPreviousResult(JavaType::class.java),
            limeParameter.typeRef.isNullable
        )
        val javaParameter =
            JavaParameter(nameRules.getName(limeParameter), javaType)
        javaParameter.comment = createComments(limeParameter)

        storeNamedResult(limeParameter, javaParameter)
        closeContext()
    }

    override fun finishBuilding(limeConstant: LimeConstant) {
        val javaConstant = JavaConstant(
            nameRules.getName(limeConstant),
            getPreviousResult(JavaType::class.java),
            getPreviousResult(JavaValue::class.java)
        )
        javaConstant.visibility = getVisibility(limeConstant)
        javaConstant.comment = createComments(limeConstant)
        addDeprecatedAnnotationIfNeeded(javaConstant)

        storeNamedResult(limeConstant, javaConstant)
        closeContext()
    }

    override fun finishBuilding(limeStruct: LimeStruct) {
        val methods = getPreviousResults(JavaMethod::class.java).map { it.shallowCopy() }
        methods.forEach { it.qualifiers.add(MethodQualifier.NATIVE) }

        val serializationBase = typeMapper.serializationBase
        val isSerializable =
            serializationBase != null && limeStruct.attributes.have(LimeAttributeType.SERIALIZABLE)

        val javaClass = JavaClass(
            name = nameRules.getName(limeStruct),
            fields = getPreviousResults(JavaField::class.java),
            methods = methods,
            constants = getPreviousResults(JavaConstant::class.java),
            isParcelable = isSerializable,
            isEquatable = limeStruct.attributes.have(LimeAttributeType.EQUATABLE),
            isImmutable = limeStruct.attributes.have(LimeAttributeType.IMMUTABLE),
            needsBuilder = limeStruct.attributes.have(JAVA, BUILDER),
            generatedConstructorComment = limeStruct.constructorComment.getFor(PLATFORM_TAG)
        )
        javaClass.visibility = getVisibility(limeStruct)
        javaClass.qualifiers.add(JavaTopLevelElement.Qualifier.FINAL)
        javaClass.javaPackage = rootPackage
        javaClass.comment = createComments(limeStruct)
        addDeprecatedAnnotationIfNeeded(javaClass)

        if (isSerializable) {
            serializationBase?.let { javaClass.parentInterfaces.add(it) }
        }

        storeNamedResult(limeStruct, javaClass)
        closeContext()
    }

    override fun finishBuilding(limeField: LimeField) {
        val fieldName = nameRules.getName(limeField)
        val javaType = typeMapper.applyNullability(
            getPreviousResult(JavaType::class.java),
            limeField.typeRef.isNullable
        )

        val defaultValue = limeField.defaultValue
        val initialValue: JavaValue
        initialValue = when {
            defaultValue != null -> valueMapper.mapValue(defaultValue, javaType)
            limeField.typeRef.isNullable -> JavaValueMapper.mapNullValue(javaType)
            else -> JavaValueMapper.mapDefaultValue(javaType)
        }

        val javaField = JavaField(fieldName, javaType, initialValue)
        javaField.visibility = getVisibility(limeField)
        javaField.comment = createComments(limeField)
        addDeprecatedAnnotationIfNeeded(javaField)

        storeNamedResult(limeField, javaField)
        closeContext()
    }

    override fun finishBuilding(limeEnumeration: LimeEnumeration) {
        val javaEnum = JavaEnum(
            nameRules.getName(limeEnumeration),
            getPreviousResults(JavaEnumItem::class.java)
        )
        javaEnum.visibility = getVisibility(limeEnumeration)
        javaEnum.javaPackage = rootPackage
        javaEnum.comment = createComments(limeEnumeration)
        addDeprecatedAnnotationIfNeeded(javaEnum)

        storeNamedResult(limeEnumeration, javaEnum)
        closeContext()
    }

    override fun finishBuilding(limeException: LimeException) {
        val limeEnumeration = limeException.errorEnum.type as LimeEnumeration
        val javaEnumType = typeMapper.mapCustomType(limeException.errorEnum.type) as JavaEnumType
        val javaException = JavaExceptionClass(nameRules.getExceptionName(limeEnumeration), javaEnumType)
        javaException.visibility = getVisibility(limeEnumeration)
        javaException.qualifiers.add(JavaTopLevelElement.Qualifier.FINAL)
        javaException.comment = createComments(limeException)
        storeNamedResult(limeException, javaException)
        closeContext()
    }

    override fun finishBuilding(limeEnumerator: LimeEnumerator) {
        val javaValue = getPreviousResultOrNull(JavaValue::class.java)
            ?: JavaValueMapper.inferNextEnumValue(
                parentContext!!.previousResults
                    .filterIsInstance<JavaEnumItem>()
                    .lastOrNull()?.value
            )
        val javaEnumItem = JavaEnumItem(
            nameRules.getName(limeEnumerator),
            javaValue
        )
        javaEnumItem.comment = createComments(limeEnumerator)
        addDeprecatedAnnotationIfNeeded(javaEnumItem)

        storeNamedResult(limeEnumerator, javaEnumItem)
        closeContext()
    }

    override fun finishBuilding(limeProperty: LimeProperty) {
        val javaType = typeMapper.applyNullability(
            getPreviousResult(JavaType::class.java),
            limeProperty.typeRef.isNullable
        )
        val qualifiers = when {
            limeProperty.isStatic -> EnumSet.of(MethodQualifier.STATIC)
            else -> EnumSet.noneOf(MethodQualifier::class.java)
        }
        val propertyComment = limeProperty.comment.getFor(PLATFORM_TAG)

        val getterComments = Comments(
            limeProperty.getter.comment.getFor(PLATFORM_TAG),
            limeProperty.getter.attributes.get(DEPRECATED, MESSAGE)
        )
        val getterMethod = JavaMethod(
            name = nameRules.getGetterName(limeProperty),
            comment = getterComments,
            visibility = getVisibility(limeProperty.getter),
            returnType = javaType,
            returnComment = propertyComment,
            qualifiers = qualifiers,
            isGetter = true
        )
        addDeprecatedAnnotationIfNeeded(getterMethod)

        storeNamedResult(limeProperty, getterMethod)
        referenceMap["${limeProperty.fullName}.get"] = getterMethod

        val limeSetter = limeProperty.setter
        if (limeSetter != null) {
            val setterParameter = JavaParameter("value", javaType)
            setterParameter.comment = Comments(propertyComment)
            val setterComments = Comments(
                limeSetter.comment.getFor(PLATFORM_TAG),
                limeSetter.attributes.get(DEPRECATED, MESSAGE)
            )
            val setterMethod = JavaMethod(
                name = nameRules.getSetterName(limeProperty),
                comment = setterComments,
                visibility = getVisibility(limeSetter),
                returnType = JavaPrimitiveType.VOID,
                parameters = listOf(setterParameter),
                qualifiers = qualifiers
            )
            addDeprecatedAnnotationIfNeeded(setterMethod)

            storeResult(setterMethod)
            referenceMap["${limeProperty.fullName}.set"] = setterMethod
        }

        closeContext()
    }

    override fun finishBuilding(limeValue: LimeValue) {
        storeResult(valueMapper.mapValue(limeValue, getPreviousResult(JavaType::class.java)))
        closeContext()
    }

    override fun finishBuilding(limeTypeRef: LimeTypeRef) {
        storeResult(typeMapper.mapType(limeTypeRef))
        closeContext()
    }

    private fun createJavaInterface(limeInterface: LimeInterface): JavaInterface {

        val javaInterface = JavaInterface(nameRules.getName(limeInterface))
        javaInterface.visibility = getVisibility(limeInterface)
        javaInterface.javaPackage = rootPackage

        javaInterface.comment = createComments(limeInterface)
        addDeprecatedAnnotationIfNeeded(javaInterface)
        javaInterface.constants.addAll(getPreviousResults(JavaConstant::class.java))
        javaInterface.enums.addAll(getPreviousResults(JavaEnum::class.java))
        javaInterface.exceptions.addAll(getPreviousResults(JavaExceptionClass::class.java))
        javaInterface.methods.addAll(getPreviousResults(JavaMethod::class.java))

        addInnerClasses(javaInterface)

        return javaInterface
    }

    private fun createJavaImplementationClass(
        limeContainer: LimeContainerWithInheritance,
        javaInterface: JavaInterface,
        extendedClass: JavaType
    ): JavaClass {

        val classMethods = getPreviousResults(JavaMethod::class.java).map { it.shallowCopy() }
        classMethods.forEach {
            it.qualifiers.add(MethodQualifier.NATIVE)
            it.visibility = JavaVisibility.PUBLIC
        }

        val javaClass = JavaClass(
            name = nameRules.getImplementationClassName(limeContainer),
            extendedClass = extendedClass,
            methods = classMethods,
            isImplClass = true,
            needsDisposer = nativeBase == extendedClass
        )
        javaClass.visibility = JavaVisibility.PACKAGE
        javaClass.javaPackage = rootPackage

        val interfaceTypeReference = JavaCustomType(javaInterface.name, javaInterface.javaPackage)
        javaClass.parentInterfaces.add(interfaceTypeReference)

        return javaClass
    }

    private fun finishBuildingClass(limeClass: LimeClass) {
        var extendedClass = nativeBase
        val parentContainer = limeClass.parent?.type as? LimeContainerWithInheritance
        if (parentContainer != null) {
            val parentClassName = when {
                parentContainer is LimeInterface ->
                    nameRules.getImplementationClassName(parentContainer)
                else -> nameRules.getName(parentContainer)
            }
            extendedClass = typeMapper.mapCustomType(parentContainer, parentClassName)
        }

        val javaClass = JavaClass(
            name = nameRules.getName(limeClass),
            extendedClass = extendedClass,
            fields = getPreviousResults(JavaField::class.java),
            isImplClass = true,
            needsDisposer = parentContainer == null,
            hasNativeEquatable = limeClass.attributes.have(LimeAttributeType.EQUATABLE) ||
                    limeClass.attributes.have(LimeAttributeType.POINTER_EQUATABLE)
        )
        javaClass.visibility = getVisibility(limeClass)
        if (limeClass.visibility == LimeVisibility.PUBLIC) {
            javaClass.qualifiers.add(JavaTopLevelElement.Qualifier.FINAL)
        }
        javaClass.javaPackage = rootPackage
        javaClass.comment = createComments(limeClass)
        addDeprecatedAnnotationIfNeeded(javaClass)

        javaClass.constants.addAll(getPreviousResults(JavaConstant::class.java))
        javaClass.methods.addAll(getPreviousResults(JavaMethod::class.java))
        javaClass.methods.forEach { it.qualifiers.add(MethodQualifier.NATIVE) }
        javaClass.enums.addAll(getPreviousResults(JavaEnum::class.java))
        javaClass.exceptions.addAll(getPreviousResults(JavaExceptionClass::class.java))

        addInnerClasses(javaClass)

        storeNamedResult(limeClass, javaClass)
    }

    private fun finishBuildingInterface(limeInterface: LimeInterface) {
        val javaInterface = createJavaInterface(limeInterface)

        var extendedClass = nativeBase
        val parentContainer = limeInterface.parent?.type as? LimeContainerWithInheritance
        if (parentContainer != null) {
            javaInterface.parentInterfaces.add(typeMapper.mapCustomType(parentContainer))
            extendedClass = typeMapper.mapCustomType(
                parentContainer,
                nameRules.getImplementationClassName(parentContainer)
            )
        }

        val javaImplementationClass =
            createJavaImplementationClass(limeInterface, javaInterface, extendedClass)

        storeNamedResult(limeInterface, javaInterface)
        storeResult(javaImplementationClass)
    }

    private fun addInnerClasses(javaTopLevelElement: JavaTopLevelElement) {
        val innerClasses = getPreviousResults(JavaClass::class.java)
            .filterNot { it.isImplClass }
        innerClasses.forEach { it.qualifiers.add(JavaTopLevelElement.Qualifier.STATIC) }
        javaTopLevelElement.innerClasses.addAll(innerClasses)
    }

    private fun getVisibility(limeElement: LimeNamedElement) =
        when {
            limeElement.visibility.isInteral -> JavaVisibility.PACKAGE
            else -> JavaVisibility.PUBLIC
        }

    private fun addDeprecatedAnnotationIfNeeded(javaElement: JavaElement) {
        if (javaElement.comment.deprecated != null) {
            javaElement.annotations.add(deprecatedAnnotation)
        }
    }

    private fun createComments(limeElement: LimeNamedElement) =
        createComments(limeElement, PLATFORM_TAG)

    companion object {
        const val PLATFORM_TAG = "Java"

        internal val deprecatedAnnotation = JavaCustomType("Deprecated")
    }
}
